### DISCUSSION TITLE
Two ingestion nodegroups are not linking "FILE" objects, but creating duplicates

### DISCUSSION THREAD

**Discussion by cuddihyge** [ge-high-assurance/RACK#444] [Q&A] [answered]
[Paul is creating this discussion from an email question]

I have a question. I have two ingestion node groups. One for the source code and one for the object code. They have a file in common. When I create the CSV file the file in common have the same identifier in both CVs. Therefore, I thought that the RACK will connect see that the identifier is the same  and therefore. It should work. However, when we ask the query that queries everything. It seems that the common file is not common. I know that the data is there and the query is correct since I create another file and used the FILTER (?file0_identifier=?file1_identifier). How should be the ingestion query so that the rack makes the connection?

**Comment by cuddihyge** [comment, marked as answer]
Great question.  I see that the wiki page, [Ingesting CSV Data](https://github.com/ge-semtk/semtk/wiki/Ingesting-CSV-Data#looking-up-uris), for this topic is correct, but not as easy-to-read as it should be.  I'll fix it.  

By default, URI's created through ingestion are created with UUID's every time.   I see from the samples that you sent that you are using this default process.  So even when a FILE has the same identifier, a copy is created with a new UUID.

You'll want to use URI look-ups.  This is a two step process that starts on the "Map Input" tab of the SPARQLgraph page.

First, click find the identifer you want to use as your unique key, and click the URI lookup button:
![image](https://user-images.githubusercontent.com/19863188/122942457-3b3ca880-d344-11eb-939b-7eabffbe113b.png)

You'll get a dialog that will allow you to pick the node, in this case **?FILE_0** is probably what you have in mind.
![image](https://user-images.githubusercontent.com/19863188/122942715-67f0c000-d344-11eb-9d2a-6d4f05c8a813.png)

After this selection is made, the URI lookup buttons on both **?objectcode_identifier** and **?FILE_0** will become highlighted.

Second, click on **?FILE_0's** URI lookup button, and choose how you would like to create the URIs:
![image](https://user-images.githubusercontent.com/19863188/122943536-1bf24b00-d345-11eb-9d00-bb59d212801c.png)

This dialog shows you that ?FILE_0 is not being used to look up any URI's, but that it is being looked up by **?obectcode_identifier**.  You have three choices for the behavior:
*  **Error if missing** -  try looking up the URI and using it.  If not found, the entire ingestion will fail.
*  **Create if missing** -  try looking up the URI and using it.  If not found, create it.
*  **Error if exists** -  try looking up the URI and if found, the entire ingestion will fail. If not found, create it.

I suspect you will use **Create if missing**, but consider these choices carefully for each ingestion nodegroup.

**Comment by dawperza** [comment]
I have a lot ingestion node groups that use _Entity_ for referencing (like polymorphism) . For example _Analysis_ can analyze _Requirements_, _Tests_, etc.

However, when I use what you mentioned before. The RACK creates an _Entity_ with the given identifier. Although, there is already a Requirement with the same identifier, which should have been selected.  

Should I create multiple node groups that ingest analysis for every type of analyzed element? Or there is a smarter way to do it?

**Reply by cuddihyge** [reply]
when you ask RACK to look up an ENTITY by identifier, it should find any subclasses with the given identifier.  I have confirmed this functionality is working in the test suite.
So either you have found a new bug or you're doing something wrong.
Please send sample nodegroups and a little data and I will investigate.

**Comment by cuddihyge** [comment]
This new topic is being moved to [Ingestion URI lookup ENTITY is not finding subclass REQUIREMENT #449](https://github.com/ge-high-assurance/RACK/discussions/449).

Let's use this current discussion for any follow-up on the original topic of linking objects across multiple ingestion runs using URI Lookup.
